File: backend/src/multimodal/image-processor.ts
import sharp from 'sharp';

export interface ImageProcessingOptions {
  resize?: { width: number; height: number };
  format?: 'jpeg' | 'png' | 'webp';
  quality?: number;
  extractText?: boolean;
  generateCaption?: boolean;
  detectObjects?: boolean;
}

export interface ProcessedImage {
  metadata: {
    width: number;
    height: number;
    format: string;
    size: number;
  };
  processed: Buffer;
  extractedText?: string;
  caption?: string;
  detectedObjects?: DetectedObject[];
}

export interface DetectedObject {
  label: string;
  confidence: number;
  bbox: { x: number; y: number; width: number; height: number };
}

export class ImageProcessor {
  async processImage(imageBuffer: Buffer, options: ImageProcessingOptions = {}): Promise<ProcessedImage> {
    // 메타데이터 추출
    const metadata = await this.extractMetadata(imageBuffer);
    
    // 이미지 처리
    let processedBuffer = imageBuffer;
    
    // 리사이징
    if (options.resize) {
      processedBuffer = await sharp(processedBuffer)
        .resize(options.resize.width, options.resize.height)
        .toBuffer();
    }
    
    // 포맷 변환
    if (options.format) {
      processedBuffer = await sharp(processedBuffer)
        .toFormat(options.format, { quality: options.quality || 85 })
        .toBuffer();
    }

    const result: ProcessedImage = {
      metadata,
      processed: processedBuffer
    };

    // OCR 텍스트 추출
    if (options.extractText) {
      result.extractedText = await this.extractText(processedBuffer);
    }

    // 이미지 캡션 생성
    if (options.generateCaption) {
      result.caption = await this.generateCaption(processedBuffer);
    }

    // 객체 검출
    if (options.detectObjects) {
      result.detectedObjects = await this.detectObjects(processedBuffer);
    }

    return result;
  }

  private async extractMetadata(imageBuffer: Buffer) {
    const image = sharp(imageBuffer);
    const metadata = await image.metadata();
    
    return {
      width: metadata.width || 0,
      height: metadata.height || 0,
      format: metadata.format || 'unknown',
      size: imageBuffer.length
    };
  }

  private async extractText(imageBuffer: Buffer): Promise<string> {
    // OCR 시뮬레이션 (실제로는 Tesseract.js 등 사용)
    console.log('Extracting text from image...');
    return 'Sample extracted text from image';
  }

  private async generateCaption(imageBuffer: Buffer): Promise<string> {
    // 이미지 캡션 생성 시뮬레이션 (실제로는 Vision 모델 사용)
    console.log('Generating image caption...');
    return 'A sample image caption describing the content';
  }

  private async detectObjects(imageBuffer: Buffer): Promise<DetectedObject[]> {
    // 객체 검출 시뮬레이션 (실제로는 YOLO, COCO 등 사용)
    console.log('Detecting objects in image...');
    
    return [
      {
        label: 'person',
        confidence: 0.95,
        bbox: { x: 100, y: 50, width: 200, height: 300 }
      },
      {
        label: 'car',
        confidence: 0.87,
        bbox: { x: 300, y: 200, width: 150, height: 100 }
      }
    ];
  }

  // 이미지 최적화
  async optimizeImage(imageBuffer: Buffer, targetSize?: number): Promise<Buffer> {
    let quality = 85;
    let optimized = imageBuffer;

    // 목표 크기가 있으면 품질 조정
    if (targetSize) {
      while (optimized.length > targetSize && quality > 10) {
        optimized = await sharp(imageBuffer)
          .jpeg({ quality })
          .toBuffer();
        quality -= 10;
      }
    }

    return optimized;
  }

  // 썸네일 생성
  async generateThumbnail(imageBuffer: Buffer, size = 150): Promise<Buffer> {
    return await sharp(imageBuffer)
      .resize(size, size, { fit: 'cover' })
      .jpeg({ quality: 80 })
      .toBuffer();
  }
}

File: backend/src/multimodal/image_processor.py
from typing import Dict, Any, Optional
import asyncio
from PIL import Image
import io
import base64


class MultiModalImageProcessor:
    def __init__(self):
        self.supported_formats = ["jpeg", "png", "webp", "gif"]

    async def process_image(
        self, image_buffer: bytes, options: Dict[str, Any]
    ) -> Dict[str, Any]:
        """이미지 처리 파이프라인"""
        results = {
            "metadata": await self.extract_metadata(image_buffer),
            "processed": None,
            "extracted_text": None,
            "caption": None,
            "objects": None,
        }

        # 이미지 처리
        processed_buffer = await self.transform_image(image_buffer, options)
        results["processed"] = base64.b64encode(processed_buffer).decode()

        # OCR 텍스트 추출
        if options.get("extract_text"):
            results["extracted_text"] = await self.extract_text(processed_buffer)

        # 캡션 생성
        if options.get("generate_caption"):
            results["caption"] = await self.generate_caption(processed_buffer)

        # 객체 검출
        if options.get("detect_objects"):
            results["objects"] = await self.detect_objects(processed_buffer)

        return results

    async def extract_metadata(self, image_buffer: bytes) -> Dict[str, Any]:
        """이미지 메타데이터 추출"""
        try:
            image = Image.open(io.BytesIO(image_buffer))
            return {
                "format": image.format,
                "mode": image.mode,
                "size": image.size,
                "width": image.width,
                "height": image.height,
            }
        except Exception:
            return {}

    async def transform_image(
        self, image_buffer: bytes, options: Dict[str, Any]
    ) -> bytes:
        """이미지 변환 처리"""
        image = Image.open(io.BytesIO(image_buffer))

        # 리사이징
        if options.get("resize"):
            size = options["resize"]
            image = image.resize((size["width"], size["height"]))

        # 포맷 변환
        format_type = options.get("format", "jpeg").upper()
        if format_type == "JPEG":
            format_type = "JPEG"

        # 품질 설정
        quality = options.get("quality", 85)

        output = io.BytesIO()
        if format_type == "JPEG":
            image = image.convert("RGB")
            image.save(output, format=format_type, quality=quality)
        else:
            image.save(output, format=format_type)

        return output.getvalue()

    async def extract_text(self, image_buffer: bytes) -> str:
        """OCR 텍스트 추출 (플레이스홀더)"""
        # 실제 구현에서는 Tesseract나 AWS Textract 사용
        return "Sample extracted text from image"

    async def generate_caption(self, image_buffer: bytes) -> str:
        """이미지 캡션 생성 (플레이스홀더)"""
        # 실제 구현에서는 BLIP, CLIP 등 사용
        return "A sample image caption"

    async def detect_objects(self, image_buffer: bytes) -> list:
        """객체 검출 (플레이스홀더)"""
        # 실제 구현에서는 YOLO, SSD 등 사용
        return [
            {"label": "person", "confidence": 0.95, "bbox": [10, 10, 100, 100]},
            {"label": "car", "confidence": 0.87, "bbox": [150, 50, 250, 150]},
        ]
